gs_chunked_io/async_collections.py

```diff
@@ -0,0 +1,125 @@
+"""
+Set-like and list-like collections to manage concurrent execution using ThreadPoolExecutor.
+"""
+import multiprocessing
+from concurrent.futures import Future, ThreadPoolExecutor, as_completed
+from typing import Any, List, Set, Callable, Generator, Optional
+
+
+class _AsyncCollection:
+    def __init__(self, executor: ThreadPoolExecutor, size: int=0):
+        self.executor = executor
+        self.size = size
+        assert 0 < self.size
+
+    def __len__(self):
+        raise NotImplementedError()
+
+    def __bool__(self) -> bool:
+        return bool(len(self))
+
+    def put(self, func, *args, **kwargs):
+        raise NotImplementedError()
+
+    def get(self) -> Any:
+        raise NotImplementedError()
+
+    def consume(self) -> Generator[Any, None, None]:
+        raise NotImplementedError()
+
+    def consume_finished(self) -> Generator[Any, None, None]:
+        raise NotImplementedError()
+
+    def _running(self) -> Set[Future]:
+        futures = getattr(self, "_futures", list())
+        return {f for f in futures if not f.done()}
+
+    def abort(self):
+        futures = getattr(self, "_futures", list())
+        for f in futures:
+            f.cancel()
+        for _ in as_completed(futures):
+            pass
+
+    def wait(self):
+        raise NotImplementedError()
+
+    def __del__(self):
+        self.abort()
+
+class AsyncSet(_AsyncCollection):
```

Comment:
It's not immediately clear to me that this is intended to be used with a GET or a PUT operation or why, except from your PR description.  Could you add a little detail into the docstring about the intended usage?

Reply:
I'm happy to provide a docstring. However, your comment implies that "get" and "put" are too closely associated with HTTP requests. I'll try to think of better names.

Reply:
`get` and `put` correspond with [queue](https://docs.python.org/3/library/queue.html#queue-objects) objects. Leaving unchanged for now.